lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,469 @@
+        return all(credentials)
+
+    def get_credentials(self, relation) -> tuple[str | None, str | None]:
+        """Return the credentials from the Juju secret.
+
+        Args:
+            relation: The relation to get the credentials for.
+
+        Returns:
+            A tuple containing the role id and secret id
+        """
+        relation = self.framework.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning("Relation is not set")
+            return None, None
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return None, None
+        if relation.app is None:
+            logger.warning("No remote application yet")
+            return None, None
+        credentials_secret_id = relation.data[relation.app].get("credentials_secret_id")
+        secret_content = self.model.get_secret(id=credentials_secret_id).get_content(refresh=True)
+        return (secret_content.get("role-id"), secret_content.get("secret-id"))
+
+
+def get_transit_key_name(relation: ops.Relation) -> str:
+    """Return the transit key name for the given relation.
+
+    Args:
+        relation: The relation to get the transit key name for.
+
+    Returns:
+        The transit key name.
+    """
+    return f"{relation.id}"
+
+
+def is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Return whether the provider data is valid.
+
+    This uses the pydantic schema to validate the data.
+
+    Args:
+        data: The data to validate.
+
+    Returns:
+        True if the data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        data = event.relation.data[event.app]
+        if is_provider_data_valid(data):
+            role_id, secret_id = self.get_credentials()
+            self.on.vault_autounseal_details_ready.emit(
+                data["address"], role_id, secret_id, data["ca_certificate"]
+            )
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent) -> None:
+        self.on.vault_autounseal_provider_removed.emit()
+
+    def get_vault_url(self) -> Optional[str]:
```

Comment:
If we need all the relation content at the same time from the tls requirer point of view, should we combine those `gets` into a single `self.vault_autounseal.get_information()`? We may also want a `self.vault_autounseal.is_information_available()`.

Reply:
I think I should be covered as the only place I'll want to check this, I'll need the details anyway if they exist, so I'll probably opt for something along the lines of 

```
details = self.requirer.get_details()
if not all(details):
     ...
```

Unless you think that's a bad approach?

That might change a little bit as I'm just considering if I should use a dataclass or namedtuple or something to clean up the return type of `get_details()`